### PR TITLE
SI-5750 scala.reflect.makro => scala.reflect.macros

### PR DESCRIPTION
**Pull request by xeno-by**

### CODE REVIEW

**Comment by xeno-by** [comment]
review by @adriaanm

**Comment by adriaanm** [comment]
excellent commit messages, thanks

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 1]
```diff
@@ -591,53 +591,34 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
...
-      // [Eugene] a heuristic to detect the REPL
-      if (global.settings.exposeEmptyPackage.value) {
-        macroLogVerbose("primary macro classloader: initializing from a REPL classloader".format(global.classPath.asURLs))
-        import scala.tools.nsc.interpreter._
-        val virtualDirectory = global.settings.outputDirs.getSingleOutput.get
-        loader = new AbstractFileClassLoader(virtualDirectory, loader) {}
-      }
+    macroLogVerbose("macro classloader: initializing from -cp: %s".format(global.classPath.asURLs))
+    val classpath = global.classPath.asURLs
+    var loader: ClassLoader = ScalaClassLoader.fromURLs(classpath, self.getClass.getClassLoader)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala]: consider avoiding the var using

```
val loader = ScalaClassLoader.fromURLs(classpath, self.getClass.getClassLoader) match {
  case loader if  global.settings.exposeEmptyPackage.value =>
    ... 
    new AbstractFileClassLoader(virtualDirectory, loader) {}
  case loader =>
    loader
```

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 2]
```diff
@@ -742,25 +723,10 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
             }
           }
 
-          val primary = loadMacroImpl(primaryClassloader)
-          primary match {
-            case Some((implObj, implMeth)) =>
+          loadMacroImpl(macroClassloader) map {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:726]: `map2 { (implObj, implMeth) =>`

**Comment by adriaanm** [comment]
> As the latest refactoring attempt has shown, the "macro" in "macro ???"
> is unnecessary and might stand in the way of macro refactorings. 

Purely for information, what exactly was the problem illustrated by the refactoring?

**Comment by adriaanm** [comment]
LGTM except for minor refactorings, which can be addressed by a follow-up commit as far as I'm concerned.

**Comment by adriaanm** [comment]
what are the steps to reproduce building the new STARR?

**Comment by cvogt** [comment]
didn't we want to make macro an identifier and call the package scala.reflection.macro?

**Comment by xeno-by** [comment]
@cvogt This didn't work: http://groups.google.com/group/scala-internals/browse_thread/thread/68888c581b005a9a.

**Comment by xeno-by** [comment]
Both commits that deploy new starrs contain all the changes needed to build those starrs. For example if you take commit [1], the starr it deploys can be obtained by applying the changes to Namers and MemberHandlers.

Do you think I should split such commits in two parts - first the compiler changes and then the new shas?

[1] https://github.com/scalamacros/kepler/commit/0a769b591d78a5e6689440c4ca285054461e35df

**Comment by xeno-by** [comment]
@adriaanm About the motivation for the first two commits. When I prototyped the syntax change from `macro foo` to `macro(foo)`, I couldn't simply change the parser and recompile, because then fast track macros stopped being recognized as macros. Hence I decided to simplify fast track stuff as much as possible.

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala (hunk 2 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:726]: value map2 is not a member of Option[(Object, java.lang.reflect.Method)]

**Comment by xeno-by** [comment]
@VladUreche how do I hide a package object from ScalaDoc? Will the facility described in https://issues.scala-lang.org/browse/SI-5533 suffice?

**Comment by VladUreche** [comment]
Yes, that's all. Did you have any problems with that?

**Comment by VladUreche** [comment]
As far as I understand, we shouldn't have package scala.reflect.makro at all in the source code, right? Then why exactly do you need to exclude it?

**Comment by xeno-by** [comment]
No problems so far, just confirming. I think it's useful to have scala.reflect.makro.Context. More details here: https://github.com/scalamacros/kepler/commit/258f9ec3459863b19300315fc5cc1d0531860cd5

**Comment by VladUreche** [comment]
Good point! Excellent use of -skip-packages!

**Comment by ijuma** [comment]
Are we sure we want to allow deprecations of code that never made to a final release? Or will this be removed before the final release and it's just to help people during the next milestone? (referring to scala.reflect.makro.Context)

**Comment by xeno-by** [comment]
I'd say this is an exceptional case. Macros gained significant popularity, even though they haven't been released yet. I think this warrants a deprecation warning when their key component is renamed.

**Comment by adriaanm** [comment]
Eugene, I agree macros created a lot of excitement with the in-crowd, but let's not exaggerate either.
Scala has >100.000 developers (yes, it boggles the mind) -- what percentage of that has written a macro?

**Comment by ijuma** [comment]
That's right. I believe there were changes that affected many more people in previous releases that did not have a deprecation cycle since they never made it into a release. If we are planning a M6, I think it makes sense to have it there. But I would remove it before the first RC. That is just my opinion though, of course.

**Comment by xeno-by** [comment]
Yeah right, but all the docs (and the blogs and the code of early adopters) that are now on the internets mention the infamous `makro`. That would definitely create confusion, even if there will be correct documentation alongside. E.g. for me the transition from sbt 0.7 to 0.10 was difficult for this very reason (of course, I don't want to compare scale of adoption for sbt and macros).

After all why not smoothen the experience of even a 0.1% of users if it only costs us a single added file in scala-reflect.jar? The scala.reflect.makro package is ignored from scaladoc, so it's not an issue for the docs.

**Comment by ijuma** [comment]
"After all why not smoothen the experience of even a 0.1% of users if it only costs us a single added file in scala-reflect.jar? The scala.reflect.makro package is ignored from scaladoc, so it's not an issue for the docs."

Two reasons:
- Many of us don't use scaladoc. I personally use IDEs and rely a lot on "Open Class", code completion and auto-import. Whenever there are duplicate names, it adds a cost. Same for people who use grep/ack. At least in the default case. Maybe configuration can be added not to show deprecated items, but it's not the case for IDEA or Eclipse by default.
- If we add one thing for each 0.1%, it adds up to a lot of things.

In any case, I suggest the decision should happen between M6 and the first RC. I would like to hear @paulp's opinion in any case.

**Comment by xeno-by** [comment]
I did overlook the IDE issue. I agree: that more or less invalidates the second argument.

**Comment by paulp** [comment]
Since @ijuma asked, my opinion is that we should never deprecate code which was never in a release.  This is what "release" means.  Everything we drag along has its price.

**Comment by jsuereth** [comment]
+1   to @paulp   Milestone releases are not releases that need deprecation.

**Comment by xeno-by** [comment]
Superceded by https://github.com/scala/scala/pull/1017

**Pull request closed without merging**